src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -39,39 +39,41 @@ public void addAll(List<T> list) {
 
     @Override
     public T get(int index) {
-        checkValidateIndex(index);
+        checkIndex(index);
         return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-        checkValidateIndex(index);
+        checkIndex(index);
         elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        checkValidateIndex(index);
-        T removed = (T) elements[index];
-        if (removed != null) {
-            System.arraycopy(elements, index + 1, elements, index, size - index - 1);
-            elements[--size] = null;
-        }
+        checkIndex(index);
+        T removed = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
         return removed;
     }
 
     @Override
     public T remove(T element) {
+        int index = index(element);
+        if (index == -1) {
+            throw new NoSuchElementException("There is no such element in the list: " + element);
+        }
+        return remove(index);
+    }
+
+    private int index(T element) {
         for (int i = 0; i < size; i++) {
-            if ((element == null && elements[i] == null) || (element != null
-                    && element.equals(elements[i]))) {
-                T removed = (T) elements[i];
-                System.arraycopy(elements, i + 1, elements, i, size - i - 1);
-                elements[--size] = null;
-                return removed;
+            if (Objects.equals(elements[i], element)) {
```

Comment:
let's not use util classes like Objects/Arrays etc in this task. Please use the solution provided by Olena